### ISSUE TITLE
the doc of Deferred.callback() doesn't tell me what will happen if the deferred has already been triggered

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6194]
|[<img alt="zooko's avatar" src="https://avatars.githubusercontent.com/u/467900?s=50" width="50" height="50">](https://github.com/zooko)| @zooko reported|
|-|-|
|Trac ID|trac#6194|
|Type|enhancement|
|Created|2012-11-22 23:45:05Z|
|Branch|https://github.com/twisted/twisted/tree/callback-docs-6194|

A colleague (David-Sarah), is proposing to add code to Tahoe-LAFS in order to work-around some unspecified limitations or problems in how Deferred handles repeated callback():

https://github.com/davidsarah/tahoe-lafs/commit/0350abb182f1fb6e4297f3c5942909a5498f5ba2#L1R46

I'd like to read some documentation about how Deferred does, or is supposed to, handle that.

Such documentation probably belongs here: http://twistedmatrix.com/documents/current/api/twisted.internet.defer.Deferred.html#callback

<details><summary>Searchable metadata</summary>

```
trac-id__6194 6194
type__enhancement enhancement
reporter__zooko zooko
priority__low low
milestone__ 
branch__branches_callback_docs_6194 branches/callback-docs-6194
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__docs docs
time__1353627905000000 1353627905000000
changetime__1353961019000000 1353961019000000
version__None None
owner__exarkun exarkun

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

(In [36449]) Branching to 'callback-docs-6194'

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_3"></a>|@glyph set owner to @exarkun|
|-|-|

Looks like a great addition to the docs, please land.

Two minor notes which you may wish to address as well:

  1. "`The object to supply as the result of this L{Deferred}.`" is obviously an improvement over what was there previously (by which I mean, of course, "nothing") but it is somewhat generally phrased.  I think a lot of redundancy and inter-linking of documentation is useful in this area.  For example, "`the object which will be passed to the first callback added to this L{Deferred} with C{addCallback}`".  Explaining what each method does in terms of _other_ methods is useful for someone who has a general idea of what is going on but is trying to piece it together by looking at reference documentation.
  2. `L{}` is superior to `C{}` if it can be made to work, since it linkifies things within the documentation and is therefore richer and more precise.  I would suggest that you might want to do something like `L{callback <Deferred.callback>}` rather than `C{callback}`.  (Maybe pydoctor should be modified so just `L{callback}` would work in that case, if it doesn't already?)

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

(In [36452]) Some more links

refs [#6194](https://github.com/twisted/twisted/issues/6194)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set status to `closed`|
|-|-|

(In [36453]) Merge callback-docs-6194

Author: exarkun
Reviewer: glyph
Fixes: [#6194](https://github.com/twisted/twisted/issues/6194)

Expand the API documentation for `Deferred.callback` and `Deferred.errback`, in
particular describing what happens if these methods are called more than once on
a single `Deferred`.
